File: typescript-effect/test/stats.test.ts
import { assertAlmostEquals, assertEquals } from "@std/assert"
import { emptyHistogram, histogram } from "../src/stats.ts"

Deno.test("histogram is empty when there are no samples", () => {
  assertEquals(histogram([]), emptyHistogram)
})

Deno.test("histogram uses the only sample for every stat", () => {
  assertEquals(histogram([12]), {
    min: 12,
    max: 12,
    average: 12,
    p50: 12,
    p90: 12,
    p99: 12,
  })
})

Deno.test("histogram nearest-rank percentiles on ten samples", () => {
  const samples = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
  const result = histogram(samples)
  assertEquals(result.min, 1)
  assertEquals(result.max, 10)
  assertAlmostEquals(result.average, 5.5)
  assertEquals(result.p50, 5)
  assertEquals(result.p90, 9)
  assertEquals(result.p99, 10)
})

File: go/stats/stats_test.go
package stats

import "testing"

func TestHistogramEmpty(t *testing.T) {
	got := From(nil)
	if got != (Histogram{}) {
		t.Fatalf("got %+v, want zeros", got)
	}
}

func TestHistogramSingleSample(t *testing.T) {
	got := From([]float64{12})
	want := Histogram{Min: 12, Max: 12, Average: 12, P50: 12, P90: 12, P99: 12}
	if got != want {
		t.Fatalf("got %+v, want %+v", got, want)
	}
}

func TestHistogramNearestRankTenSamples(t *testing.T) {
	got := From([]float64{1, 2, 3, 4, 5, 6, 7, 8, 9, 10})
	if got.Min != 1 || got.Max != 10 {
		t.Fatalf("min/max %+v", got)
	}
	if got.Average != 5.5 {
		t.Fatalf("average %v", got.Average)
	}
	if got.P50 != 5 || got.P90 != 9 || got.P99 != 10 {
		t.Fatalf("percentiles %+v", got)
	}
}
